WearTilesKotlin/app/src/main/java/com/example/wear/tiles/messaging/MessagingTileService.kt

```diff
@@ -35,77 +35,87 @@ import kotlinx.coroutines.flow.map
- * This is a demo tile only, so the buttons don't actually work.
+ * Creates a Messaging tile, showing up to 4 contacts, an icon button and compact chip.
  *
- * The main function, [onTileRequest], is triggered when the system calls for a tile and implements
- * ListenableFuture which allows the Tile to be returned asynchronously.
+ * It extends [CoroutinesTileService], a Coroutine-friendly wrapper around
+ * [androidx.wear.tiles.TileService], and implements [tileRequest] and [resourcesRequest].
  *
- * Resources are provided with the [onResourcesRequest] method, which is triggered when the tile
- * uses an Image.
+ * The main function, [tileRequest], is triggered when the system calls for a tile. Resources are
+ * provided with the [resourcesRequest] method, which is triggered when the tile uses an Image.
  */
 class MessagingTileService : CoroutinesTileService() {
-    private lateinit var tileStateFlow: StateFlow<MessagingTileState?>
-    private lateinit var renderer: MessagingTileRenderer
     private lateinit var repo: MessagingRepo
     private lateinit var imageLoader: ImageLoader
+    private lateinit var renderer: MessagingTileRenderer
+    private lateinit var tileStateFlow: StateFlow<MessagingTileState?>
 
     override fun onCreate() {
         super.onCreate()
 
-        val appContainer = (application as TilesApplication).appContainer
-        renderer = appContainer.renderer
-        repo = appContainer.repo
-        imageLoader = ImageLoader(this)
+        repo = MessagingRepo(this)
+        imageLoader = Coil.imageLoader(this)
+        renderer = MessagingTileRenderer(this)
 
         tileStateFlow = repo.getFavoriteContacts()
-            .map {
-                buildState(it)
-            }
+            .map { contacts -> MessagingTileState(contacts) }
             .stateIn(
                 lifecycleScope,
                 started = SharingStarted.WhileSubscribed(5000),
                 initialValue = null
             )
     }
 
-    private fun buildState(contacts: List<Contact>): MessagingTileState {
-        return MessagingTileState(contacts)
-    }
-
+    /**
+     * Read the latest data, delegating to the [MessagingTileRenderer] which creates a layout to
+     * which it binds the state.
+     */
     override suspend fun tileRequest(requestParams: TileRequest): Tile {
-        val tileState = readTileState()
-
+        val tileState: MessagingTileState = latestTileState()
         return renderer.renderTimeline(tileState, requestParams)
     }
 
-    private suspend fun readTileState(): MessagingTileState {
+    /**
+     * Reads the latest state from the flow, and updates the data if there isn't any.
+     */
+    private suspend fun latestTileState(): MessagingTileState {
         var tileState = tileStateFlow.filterNotNull().first()
 
         if (tileState.contacts.isEmpty()) {
-            updateContacts()
+            refreshData()
```

Comment:
TIOLI: As a sample, maybe this deserves a comment that we only have this here because the demo doesn't have a real backend? In a real app, app, complications, tiles would all share this common data source, and data refreshes wouldn't be triggered by a tile being installed.

Reply:
Actually saw the comment below for refreshData, maybe could slightly expand that and refer to that from here.

Reply:
What if the tile is the first thing that's added (before the overlay is opened)? In that case, I would expect the tile to be responsible for fetching the data initially.

Reply:
Yep.  But it also has a much tighter window (x seconds) to complete, so can be problematic then.